tools/src/main/java/org/apache/rocketmq/tools/command/message/ExportMessageCommand.java

```diff
@@ -0,0 +1,242 @@
+            if (StringUtils.isBlank(brokerName) && queueId != -1) {
+                throw new SubCommandException("Please set the brokerName before queueId!");
+            }
+            String directory =
+                !commandLine.hasOption('d') ? DEFAULT_EXPORT_DIRECTORY : commandLine.getOptionValue('s').trim();
+            String bodyFormat =
+                !commandLine.hasOption('f') ? "base64" : commandLine.getOptionValue('f').trim();
+
+            defaultMQPullConsumer.start();
+            Set<MessageQueue> messageQueues = Collections.emptySet();
+            if (StringUtils.isNotBlank(brokerName) && queueId != -1) {
+                messageQueues = Sets.newHashSet(new MessageQueue(topic, brokerName, queueId));
+            } else if (StringUtils.isNotBlank(brokerName) && queueId == -1) {
+                messageQueues = defaultMQPullConsumer.fetchSubscribeMessageQueues(topic).stream().filter(e -> e.getBrokerName().equals(brokerName)).collect(Collectors.toSet());
+            } else if (StringUtils.isBlank(brokerName) && queueId == -1) {
+                messageQueues = defaultMQPullConsumer.fetchSubscribeMessageQueues(topic);
+            }
+
+            String topicDir = directory + File.separator + topic;
+            FileUtils.forceDeleteOnExit(new File(topicDir));
+            FileUtils.forceMkdirParent(new File(topicDir));
+            for (MessageQueue mq : messageQueues) {
+                String brokerDir = topicDir + File.separator + mq.getBrokerName();
+                String queueFilepath = brokerDir + File.separator + mq.getQueueId();
+                File brokerDirFile = new File(brokerDir);
+                FileUtils.forceMkdir(brokerDirFile);
+                File queueFile = new File(queueFilepath);
+                System.out.printf("exporrt queueFile queueFilepath=%s%n", queueFile.getAbsoluteFile());
+
+                long minOffset = defaultMQPullConsumer.minOffset(mq);
+                long maxOffset = defaultMQPullConsumer.maxOffset(mq);
+
+                if (commandLine.hasOption('b')) {
+                    String timestampStr = commandLine.getOptionValue('b').trim();
+                    long timeValue = timestampFormat(timestampStr);
+                    minOffset = defaultMQPullConsumer.searchOffset(mq, timeValue);
+                }
+
+                if (commandLine.hasOption('e')) {
+                    String timestampStr = commandLine.getOptionValue('e').trim();
+                    long timeValue = timestampFormat(timestampStr);
+                    maxOffset = defaultMQPullConsumer.searchOffset(mq, timeValue);
+                }
+
+                System.out.printf("export %s minOffset=%s, maxOffset=%s%n", minOffset, maxOffset, mq);
+                try (BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(queueFile), charsetName))) {
+                    READQ:
+                    for (long offset = minOffset; offset < maxOffset; ) {
```

Comment:
It is recommended to be asynchronous, and to support breakpoint resume from commited offset.